### PR TITLE
898 enhancement add a new multipart parser inside starlite

### PR DESCRIPTION
**Pull request by Goldziher**
# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.md`?
- [x] Have you got 100% test coverage on new code?
- [x] Have you updated the prose documentation?
- [x] Have you updated the reference documentation?

rps_post-body

This PR adds a highly optimized multipart form parser logic, vendored from Sanic and reworked. It shows huge performance boosts over the older approach we used, which was a statemachine based solution. I also updated our handling of urlencoded query parameters, which shows signficiant boost as well, and added cleanup to ensure temporary files are properly closed. 

Closes both #904 and #898

### CODE REVIEW

File: starlite/constants.py [hunk 1]
```diff
@@ -12,3 +12,4 @@
 SCOPE_STATE_DEPENDENCY_CACHE = "dependency_cache"
 SCOPE_STATE_NAMESPACE = "__starlite__"
 SCOPE_STATE_RESPONSE_COMPRESSED = "response_compressed"
+ONE_MEGA_BYTE: int = 1024 * 1024
```
Comment (cofin) [inline starlite/constants.py:15]:
```suggestion
ONE_MEGABYTE: int = 1024 * 1024
```

File: starlite/datastructures/upload_file.py [hunk 2]
```diff
@@ -1,16 +1,107 @@
...
 
+    __slots__ = ("filename", "file", "content_type", "headers")
+
+    def __init__(
+        self,
+        content_type: str,
+        filename: str,
+        file_data: Optional[bytes] = None,
+        headers: Optional[Dict[str, str]] = None,
+        max_spool_size: int = ONE_MEGA_BYTE,
```
Comment (cofin) [inline starlite/datastructures/upload_file.py:24]:
```suggestion
        max_spool_size: int = ONE_MEGABYTE,
```

File: starlite/response/file.py [hunk 3]
```diff
@@ -15,6 +15,7 @@
 from urllib.parse import quote
 from zlib import adler32
 
+from starlite.constants import ONE_MEGA_BYTE
```
Comment (cofin) [inline starlite/response/file.py:18]:
```suggestion
from starlite.constants import ONE_MEGABYTE
```

**Review by cofin — APPROVED** [review]
Looks great to me.  The testing on my current application showed really great results when using this branch as well.

**Comment by defnull** [comment]
Wait, the new parser holds the entire request body in memory as a single byte string? Am I missing something? No wonder this is faster, but also useless for large file uploads. Interesting choice.

**Pull request merged**